updated buildkite pipeline generation

metaflow opened the pull request:
- fixed build for linux (clang was missing)

- removed /monolithic-.. from build directory - it does not add anything and makes path longer for windows which is not great;

- added env-based configuration to control cache and agent targeting;

- print (s)ccache stats to file not to pullute normal log.

aeubanks commented on .ci/monolithic-linux.sh:
```diff
@@ -17,14 +17,19 @@ set -ex
 ccache --zero-stats
-ccache --show-config
+
+if [[ -n "${CLEAR_CACHE:-}" ]]; then
```
how can you set this?

metaflow replied:
you can do that when creating a build in buildkite UI or triggering through API with "environment" varaibles. It's not available for "normal" users

aeubanks commented on .ci/generate-buildkite-pipeline-scheduled:
```diff
@@ -28,42 +40,42 @@ steps:
-      - 'C:\\BuildTools\\Common7\\Tools\\VsDevCmd.bat -arch=amd64 -host_arch=amd64'
-      - 'bash .ci/monolithic-windows.sh "clang-tools-extra;flang;libclc;lld;llvm;mlir;polly;pstl" "check-all"'
+    - C:\\BuildTools\\Common7\\Tools\\VsDevCmd.bat -arch=amd64 -host_arch=amd64
+    - bash .ci/monolithic-windows.sh "clang-tools-extra;flang;libclc;lld;llvm;mlir;polly;pstl;clang" "check-all"
```
make alphabetical just to be consistent and look nice

metaflow replied:
Done.

aeubanks commented on .ci/generate-buildkite-pipeline-scheduled:
```diff
@@ -28,42 +40,42 @@ steps:
-    agents:
-      queue: 'linux'
+    - artifacts/**/*
+    - '*_result.json'
```
quoting is inconsistent, I believe all of these don't need quotes
(other files as well)

metaflow replied:
Done.

aeubanks reviewed (commented):
the commit title doesn't make sense, can you make it something like "Various fixes to buildkite pipeline generation"

metaflow committed:
update buildkite pipeline generation

- fixed build for linux (clang was missing)

- removed /monolithic-.. from build directory - it does not add anything
  and makes path longer for windows which is not great;

- added env-based configuration to control cache and agent targeting;

- print (s)ccache stats to file not to pullute normal log.

metaflow commented:
Test build https://buildkite.com/llvm-project/experimental-premerge-checks/builds/60

joker-eph commented:
This seems broken: https://buildkite.com/llvm-project/github-pull-requests/builds/1350#018a8848-f70e-47f4-8aec-0bb48c49a7ca

```
$ trap 'kill -- $' INT TERM QUIT; .ci/generate-buildkite-pipeline-premerge \| buildkite-agent pipeline upload
  | 2023-09-12 07:26:50 INFO   Reading pipeline config from STDIN
  | 2023-09-12 07:26:50 INFO   Updating BUILDKITE_COMMIT to "6de56a7dc4ff114eb36f42f1dbe8c63e557f061b"
  | Files modified:
  | mlir/include/mlir/Dialect/Affine/IR/AffineOps.h
  | mlir/include/mlir/Dialect/Affine/IR/AffineOps.td
  | mlir/include/mlir/IR/SymbolInterfaces.td
  | mlir/lib/Dialect/Affine/IR/AffineOps.cpp
  | mlir/test/Dialect/Affine/invalid.mlir
  | mlir/test/Dialect/Affine/load-store-invalid.mlir
  | Directories modified:
  | mlir
  | 2023-09-12 07:26:50 FATAL  Pipeline parsing of "(stdin)" failed: line 5: did not find expected alphabetic or numeric character
  | 🚨 Error: The command exited with status 1
```


This seems to match the ` - '*_result.json'`  replaced with ` - *_result.json`

metaflow commented:
thanks for the heads up @joker-eph ! I should be fixed now but you need to rebase your PR to see the change. I guess noop commit update should do

metaflow commented:
ah you've merged it already :smile:

ldionne commented:
@metaflow The reason for not adding clang to the monolithic pipeline is that clang now has its own pipeline like libc++. After your patch we will be building and testing clang twice for no reason. Can you remove it again?

metaflow commented:
@ldionne  w/o clang flang will not compile. Do you also check flang and all dependencies of clang in clang pipeline?

ldionne commented:
I think it's fine to add `clang` to the set of projects in `LLVM_ENABLE_PROJECTS` for dependency purposes, but we could avoid calling `check-clang`. I guess it's moot cause we're using `check-all` right now.

We don't check `flang` in the Clang pipeline because the Clang pipeline's primary responsibility should be to test Clang, not other downstream projects.

joker-eph commented:
> I think it's fine to add clang to the set of projects in LLVM_ENABLE_PROJECTS for dependency purposes, but we could avoid calling check-clang. I guess it's moot cause we're using check-all right now.


FYI, the script for premerge generates invocations like:

`'./.ci/monolithic-linux.sh "clang;flang;llvm;mlir" "check-flang check-mlir"'`

So even though `clang` is enabled, it shouldn't be tested I think? (we don't want to test LLVM either when it isn't touched)

ldionne commented:
The scheduled build script does call `check-all`, but you're right about the script that generates pre-commit CI pipelines.

mtvec commented on .ci/monolithic-windows.sh:
```diff
@@ -45,4 +51,4 @@ cmake -S ${MONOREPO_ROOT}/llvm -B ${BUILD_DIR} \
 
 echo "--- ninja"
-ninja -C ${BUILD_DIR} ${targets}
+ninja -C "${BUILD_DIR}" "${targets}"
```
Could it be that adding quotes around `targets` breaks Windows build? For example [here](https://buildkite.com/llvm-project/github-pull-requests/builds/1606#018a8da6-0dda-4d24-a1cf-971e3450e4f8):

```
+ ninja -C C:/ws/src/build 'check-clang check-clang-tools check-flang check-lld check-llvm check-mlir check-polly'
ninja: error: unknown target 'check-clang check-clang-tools check-flang check-lld check-llvm check-mlir check-polly'
```

metaflow replied:
on it

metaflow replied:
fixed in #66192 , thanks for the headsup

mtvec replied:
Thanks for the quick fix!
